sql3/planner/expressionagg.go

```diff
@@ -943,6 +944,311 @@ func (n *percentilePlanExpression) WithChildren(children ...types.PlanExpression
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// corrPlanExpression handles CORR() - implement correlation coefficient
+type corrPlanExpression struct {
+	arg1           types.PlanExpression
+	arg2           types.PlanExpression
+	returnDataType parser.ExprDataType
+}
+
+var _ types.Aggregable = (*corrPlanExpression)(nil)
+
+func newCorrPlanExpression(arg1 types.PlanExpression, arg2 types.PlanExpression, returnDataType parser.ExprDataType) *corrPlanExpression {
+	return &corrPlanExpression{
+		arg1:           arg1,
+		arg2:           arg2,
+		returnDataType: returnDataType,
+	}
+}
+
+func (n *corrPlanExpression) Evaluate(currentRow []interface{}) (interface{}, error) {
+	return nil, sql3.NewErrInternalf("this should never be called")
+}
+
+func (n *corrPlanExpression) NewBuffer() (types.AggregationBuffer, error) {
+	return NewAggCorrBuffer(n), nil
+}
+
+func (n *corrPlanExpression) FirstChildExpr() types.PlanExpression {
+	return n.arg1
+}
+
+func (n *corrPlanExpression) Type() parser.ExprDataType {
+	return n.returnDataType
+}
+
+func (n *corrPlanExpression) String() string {
+	return fmt.Sprintf("corr(%s, %s)", n.arg1.String(), n.arg2.String())
+}
+
+func (n *corrPlanExpression) Plan() map[string]interface{} {
+	result := make(map[string]interface{})
+	result["_expr"] = fmt.Sprintf("%T", n)
+	result["description"] = n.String()
+	result["dataType"] = n.Type().TypeDescription()
+	result["arg1"] = n.arg1.Plan()
+	result["arg2"] = n.arg2.Plan()
+	return result
+}
+
+func (n *corrPlanExpression) Children() []types.PlanExpression {
+	return []types.PlanExpression{
+		n.arg1,
+		n.arg2,
+	}
+}
+
+func (n *corrPlanExpression) WithChildren(children ...types.PlanExpression) (types.PlanExpression, error) {
+	if len(children) != 2 {
+		return nil, sql3.NewErrInternalf("unexpected number of children '%d'", len(children))
+	}
+	return newCorrPlanExpression(children[0], children[1], n.returnDataType), nil
+}
+
+// aggregator for VAR()
+type aggregateVar struct {
+	expr *varPlanExpression
+
+	// to calculate mean
+	n   int64
+	sum float64
+
+	// we need to hang on to the values
+	// TODO(pok) - will need to spill these to disk for big result sets
+	values []float64
+}
+
+func NewAggVarBuffer(child *varPlanExpression) *aggregateVar {
+	return &aggregateVar{
+		expr:   child,
+		values: make([]float64, 0),
+	}
+}
+
+func (m *aggregateVar) Update(ctx context.Context, row types.Row) error {
+	v, err := m.expr.arg.Evaluate(row)
+	if err != nil {
+		return err
+	}
+
+	// skip if nil
+	if v == nil {
+		return nil
+	}
+
+	var val float64
+
+	switch dataType := m.expr.arg.Type().(type) {
+	case *parser.DataTypeDecimal:
+		thisVal, ok := v.(pql.Decimal)
+		if !ok {
+			return sql3.NewErrInternalf("unexpected type conversion '%T'", v)
+		}
+
+		val = thisVal.Float64()
+
+	case *parser.DataTypeID:
+		thisVal, ok := v.(int64)
+		if !ok {
+			return sql3.NewErrInternalf("unexpected type conversion '%T'", v)
+		}
+
```

Comment:
I think we're a ways out from naturally-generated IDs out of range, but we have historically at least in principle supported pretty close to a full 63-bit range of ID values. I feel like we should probably support int64 values which aren't converted to float64 if at all possible.

Reply:
can't do variance without floats, so everything with these new types is float.